### PR TITLE
Add UnaryExistsOpPatcher

### LINKED ISSUE #126: New patcher for UnaryExistsOp
First of, thanks so much for writing all this, it's super useful!

I'm using the patchers branch with the decaf-parser and alot of the code I'm currently converting has lots of
```
if data.promocode?
  stuff()

if data.blah?
  something_else()
```

to convert this it looks like a patcher needs to be written for  `UnaryExistsOp` and I'm happy to write it

I've had a look through the existing patchers but something isn't quite clicking.. are there any tips or pointers you could give, or at least suggest a good patcher to start from.

Thanks!

### PR DESCRIPTION
**Pull request by jeffchan**
Closes #126

First stab at this, many thanks to https://github.com/decaffeinate/decaffeinate/blob/master/CONTRIBUTING.md (very helpful)

Any feedback?

### CODE REVIEW

**Comment by jeffchan** [comment]
Not sure why the Travis tests aren't running.

**Comment by eventualbuddha** [comment]
I forgot to set up the Travis tests again after moving to the decaffeinate org. They should be working again.

File: src/patchers/UnaryExistsOpPatcher.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    // Only add parentheses if necessary
+    if (node.parentNode && node.parentNode.type !== 'Block') {
+        this.insertBefore('(');
+        this.insertAfter(')');
+    }
+
+    if (nodeExpression && nodeExpression.type === 'Identifier') {
+      this.overwrite(this.start, this.end, `typeof ${nodeExpression.raw} !== 'undefined' && ${nodeExpression.raw} !== null`);
+    } else {
+      this.overwrite(this.expression.after, this.end, ' != null');
```
Comment (eventualbuddha) [inline src/patchers/UnaryExistsOpPatcher.js]: I've started including comments here to explain more clearly what's going on. Here I'd put this:

```
// `a?` → `a != null`
//   ^      ^^^^^^^^
```

And same for the `Identifier` case.

File: src/patchers/UnaryExistsOpPatcher.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import UnaryOpPatcher from './UnaryOpPatcher.js';
+
+export default class UnaryExistsOpPatcher extends UnaryOpPatcher {
+  patchAsExpression() {
+    const node = this.node;
+    const nodeExpression = node.expression;
+
+    // Only add parentheses if necessary
+    if (node.parentNode && node.parentNode.type !== 'Block') {
```
Comment (eventualbuddha) [inline src/patchers/UnaryExistsOpPatcher.js]: This is pretty much trying to replicate the statement vs. expression thing. What I'd probably do is to put everything from `patchAsExpression` into `patchAsStatement` and then call the latter from the former like so:

```js
patchAsExpression() {
  this.insertBefore('(');
  this.patchAsStatement();
  this.insertAfter(')');
}
```

File: src/patchers/UnaryExistsOpPatcher.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        this.insertBefore('(');
+        this.insertAfter(')');
+    }
+
+    if (nodeExpression && nodeExpression.type === 'Identifier') {
```
Comment (eventualbuddha) [inline src/patchers/UnaryExistsOpPatcher.js]: For extra credit, you should handle negation a la [`EqualityPatcher`](https://github.com/decaffeinate/decaffeinate/blob/master/src/patchers/EqualityPatcher.js). You should add tests like this:

```js
check(`
  unless a?
    b
`, `
  if (typeof a === 'undefined' || a === null) {
    b;
  }
`);

check(`
  unless a.b?
    b
`, `
  if (a.b == null) {
    b;
  }
`);
```

Comment (jeffchan) [inline src/patchers/UnaryExistsOpPatcher.js]: Ok, will attempt this :smile:

**Comment by jeffchan** [comment]
@eventualbuddha  -- getting stuck on this. I would've expected `isUnless` to be true. What am I missing? (Or perhaps I'm going down the wrong path)

```js
{ type: 'Conditional',
  line: 1,
  column: 1,
  range: [ 0, 13 ],
  isUnless: false,
  condition:
   { type: 'UnaryExistsOp',
     line: 1,
     column: 8,
     range: [ 7, 9 ],
     expression:
      { type: 'Identifier',
        line: 1,
        column: 8,
        range: [Object],
        data: 'a',
        raw: 'a',
        parentNode: [Circular],
        scope: Scope {} },
     raw: 'a?',
     parentNode: [Circular],
     scope: Scope {} },
  consequent:
   { type: 'Block',
     line: 2,
     column: 3,
     range: [ 12, 13 ],
     statements: [ [Object] ],
     raw: 'b',
     inline: false,
     parentNode: [Circular],
     scope: Scope {} },
  alternate: null,
  raw: 'unless a?\n  b',
  parentNode:
   { type: 'Block',
     line: 1,
     column: 1,
     range: [ 0, 13 ],
     statements: [ [Circular] ],
     raw: 'unless a?\n  b',
     parentNode:
      { type: 'Program',
        line: 1,
        column: 1,
        range: [Object],
        body: [Circular],
        raw: 'unless a?\n  b',
        scope: Scope {} },
     scope: Scope {} },
  scope: Scope {} }
```

**Comment by eventualbuddha** [comment]
@jeffchan that does seem incorrect. Perhaps I did something wrong in decaffeinate-parser. I'll check. If decaffeinate-parser is doing its thing correctly you shouldn't have to look at the `Conditional` node at all, it'll call `negate()` on your patcher for you.

**Comment by jeffchan** [comment]
Yeah, for what it's worth, I don't think `ConditionalPatcher` is visited in this case.

**Comment by eventualbuddha** [comment]
@jeffchan should be fixed by 1cbae89eeec80dc90566adf3f9fe2ab628d151ea and https://github.com/decaffeinate/decaffeinate-parser/commit/3c0eb9d6e9071c69b9205be6c51870358e1d6b23.

File: test/unary_operator_test.js [hunk 4]
```diff
@@ -40,17 +40,38 @@ describe('unary operators', () => {
...
+      if ((typeof a === 'undefined' || a === null) || !c) {
+        b;
+      }
+    `);
+
+    check(`
+      unless a.b?
+        b
+    `, `
+      if ((a.b == null)) {
```
Comment (jeffchan) [inline test/unary_operator_test.js]: The double parens is the last piece that I'm struggling with. Do I need to detect that the conditional has no surrounding parens, then I can assume that `ConditionalPatcher` will take care of it, so I should not insert the parens? What about compound expressions?

Comment (eventualbuddha) [inline test/unary_operator_test.js]: Ah. That's probably beyond the scope of this PR. Let's just test for the
double parens. What is the question about compound expressions?

Comment (jeffchan) [inline test/unary_operator_test.js]: Ah, nevermind. I was just rambling about how the solution I had in my head probably wouldn't work for compound expressions :stuck_out_tongue:

**Comment by jeffchan** [comment]
@eventualbuddha Ready for another look!

**Comment by eventualbuddha** [comment]
Thanks!

**Pull request merged**